core/src/main/java/nl/nn/adapterframework/pgp/PGPAction.java

```diff
@@ -0,0 +1,111 @@
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+package nl.nn.adapterframework.pgp;
+
+import name.neuhalfen.projects.crypto.bouncycastle.openpgp.keys.callbacks.KeyringConfigCallback;
+import name.neuhalfen.projects.crypto.bouncycastle.openpgp.keys.callbacks.KeyringConfigCallbacks;
+import name.neuhalfen.projects.crypto.bouncycastle.openpgp.keys.keyrings.InMemoryKeyring;
+import name.neuhalfen.projects.crypto.bouncycastle.openpgp.keys.keyrings.KeyringConfigs;
+import nl.nn.adapterframework.configuration.ConfigurationException;
+import org.apache.commons.io.IOUtils;
+import org.bouncycastle.openpgp.PGPException;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+/**
+ * This is an abstraction of general pgp actions
+ * such as encryption, verification, etc.
+ * to be used for {@link nl.nn.adapterframework.pipes.PGPPipe}
+ *
+ * @author Murat Kaan Meral
+ */
+public abstract class PGPAction {
+
+	protected InMemoryKeyring keyringConfig;
+	private String[] publicKeys;
+	private String secretKey, secretPassword;
+
+	/**
+	 * A general constructor. It checks if any of the vars are null.
+	 *
+	 * @param publicKey      Array of path strings that point to public keys.
+	 * @param secretKey      String that contains the path to the secret key.
+	 * @param secretPassword String that contains the password for the secret key.
+	 * @param vars           Objects that should not be null for the action to be performed.
+	 * @throws ConfigurationException When any of the given vars is null.
+	 */
+	PGPAction(String[] publicKey, String secretKey, String secretPassword, Object... vars) throws ConfigurationException {
+		verifyNotNull(vars);
+		this.publicKeys = publicKey;
+		this.secretPassword = secretPassword;
+		this.secretKey = secretKey;
+	}
+
+	/**
+	 * Generates a keyring configuration with public keys and the private key.
+	 *
+	 * @throws ConfigurationException When the files do not exist, or unexpected PGP exception has occurred.
+	 */
+	public void configure() throws ConfigurationException {
+		try {
+			// Create configuration
+			KeyringConfigCallback callback = KeyringConfigCallbacks.withUnprotectedKeys();
+			if (secretPassword != null)
+				callback = KeyringConfigCallbacks.withPassword(secretPassword);
+
+			keyringConfig = KeyringConfigs.forGpgExportedKeys(callback);
+
+			// Add public keys
+			if (publicKeys != null) {
+				for (String s : publicKeys) {
+					keyringConfig.addPublicKey(IOUtils.toByteArray(new FileInputStream(s)));
```

Comment:
Can you use ClassUtils.getResourceURL ? it is very unlikely that the resource will be on the local filesystem. this generic method will use the appropiate classloader to retrieve the url

Reply:
Is it also possible to retrieve these certificates from the JVM keystore? With a property to specify the name of the keystore to retrieve the certificate from

Reply:
For getResourceURL should I use this.getClass().getClassLoader() as classloader or something else? 
Bouncy Castle and Java Keystore API are not compatible. If we need it, I could create an adapter to make them compatible. However, that would require rebuilding the keys each time. See [this url.](https://stackoverflow.com/questions/35928359/how-can-i-store-a-bouncy-castle-pgp-key-in-a-java-keystore)

Reply:
> For getResourceURL should I use this.getClass().getClassLoader() as classloader or something else?
Have a look at ClassUtils.getResourceURL

> Bouncy Castle and Java Keystore API are not compatible. If we need it, I could create an adapter to make them compatible. However, that would require rebuilding the keys each time. See [this url.](https://stackoverflow.com/questions/35928359/how-can-i-store-a-bouncy-castle-pgp-key-in-a-java-keystore)
Fair enough, was a nice to have in my eyes!